File: 51.n-queens/solution.cpp
class Solution {
public:
	vector<vector<string>> solveNQueens(int n) {
		vector<vector<string>> result;
		int *answer = new int[n];
		int **p = new int*[n];
		for (int i = 0; i < n; i++)
		{
			p[i] = new int[n];
			memset(p[i], 0, n * sizeof(int));
		}
		solveNQueens(result, answer, p, n, 0);
		for (int i = 0; i < n; i++)
			delete[] p[i];
		delete[] p;
		return result;
	}
private:
	void solveNQueens(vector<vector<string>> &result, int *answer, int **p, int n, int j)
	{
		if (j == n)
		{
			result.emplace_back(n, string(n, '.'));
			for (int i = 0; i < n; i++)
				result.back()[i][answer[i]] = 'Q';
			return;
		}
		for (int i = 0; i < n; i++)
		{
			if (!p[j][i])
			{
				answer[j] = i;
				for (int k = j + 1; k < n; k++)
				{
					p[k][i]++;
					int i1 = k - j + i;
					int i2 = j - k + i;
					if (i1 < n)
						p[k][i1]++;
					if (i2 >= 0)
						p[k][i2]++;
				}
				solveNQueens(result, answer, p, n, j + 1);
				for (int k = j + 1; k < n; k++)
				{
					p[k][i]--;
					int i1 = k - j + i;
					int i2 = j - k + i;
					if (i1 < n)
						p[k][i1]--;
					if (i2 >= 0)
						p[k][i2]--;
				}
			}
		}
	}
};

File: 51.n-queens/solution.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        sum = new boolean[n * 2 - 1];
        diff = new boolean[n * 2 - 1];
        line = new boolean[n];
        List<List<String>> result = new ArrayList<>();
        solveNQueens(result, n, new ArrayList<>());
        return result;
    }

    private void solveNQueens(List<List<String>> result, int n, List<Integer> list) {
        int j = list.size();
        if (j == n) {
            List<String> strings = new ArrayList<>();
            for (int i : list) {
                char[] ch = new char[n];
                for (int x = 0; x < n; x++)
                    ch[x] = x == i ? 'Q' : '.';
                strings.add(new String(ch));
            }
            result.add(strings);
            return;
        }
        list.add(0);
        for (int i = 0; i < n; i++) {
            if (line[i]) continue;
            if (sum[i + j]) continue;
            int diffVal = i - j + n - 1;
            if (diff[diffVal]) continue;
            line[i] = true;
            sum[i + j] = true;
            diff[diffVal] = true;
            list.set(j, i);
            solveNQueens(result, n, list);
            line[i] = false;
            sum[i + j] = false;
            diff[diffVal] = false;
        }
        list.remove(j);
    }

    private boolean[] sum;
    private boolean[] diff;
    private boolean[] line;
}
